### ISSUE TITLE
cannot use Vue i18n in slot templates

### ISSUE THREAD

**Issue by arshaw** [fullcalendar/fullcalendar-vue#105]
Reported by @AbrahamBrookes in #14:

> It seems that using the new slot feature breaks reference to my root vue app? I have a component that sits inside the eventContent slot, and I can no longer access my root vue instance (in particular, my vuex store) from that component:

```
		<full-calendar 
			:options="calOptions"
		>
			<template v-slot:eventContent="event" >
				<planner-period
					:event="event"
				></planner-period>
			</template>

		</full-calendar>
```

> Inside the planner-period component when I try to access my store data using this.$store.state I get this.$store is undefined. Is the calendar creating its own entirely new vue app?

**Comment by arshaw** [comment]
@AbrahamBrookes, Would you be able to post a [runnable, stripped-down demonstration of the bug](https://fullcalendar.io/reduced-test-cases)? Would really appreciate it because the time saved reproducing will be time spent fixing. I'm closing this issue until then.

**Comment by kgilden** [comment]
@arshaw I have managed to reproduce this. Please see https://repl.it/@kgilden/fullcalendar-vue-105#components.js.

Would you consider reopening this issue?

**Comment by slavabelb** [comment]
I've got the same problem here.
The template of component looks like
    <FullCalendar :options="calendarOptions" class="kk-calendar">
      <template #eventContent="arg">
        <b>{{ arg.timeText }}</b>
        <i>{{ arg.event.title }}</i> 
      </template>
    </FullCalendar>
But body of every event is empty.
I've checked the dom representation and <div class="fc-event-main"> have no childs

**Comment by kgilden** [comment]
@slavabelb this seems unrelated to this specific issue. Your slot is not accessing store or i18n inside a component. As such, it's likely better to create a new issue rather than do it here.

**Comment by arshaw** [comment]
sorry. i'll take a look at this again!

**Comment by arshaw** [comment]
@kgilden, from an internal codebase point of view, this won't work because each slot is a unique instantiation of a Vue object that is not properly associated with the root Vue instance you are creating. See:
https://github.com/fullcalendar/fullcalendar-vue/blob/v5.3.0/src/custom-content-type.ts#L52

Thus, it won't inherit your Vue i18n configuration.

I'm wondering, as a workaround, could you somehow initialize the i18n plugin globally, using `Vue.use(VueI8n)` (like [here](https://kazupon.github.io/vue-i18n/installation.html#yarn))?

**Comment by kgilden** [comment]
> this won't work because each slot is a unique instantiation of a Vue object that is not properly associated with the root Vue instance you are creating.

@arshaw okay, this sounds like a logical explanation as to why i18n is undefined. However, let me clarify that i18n is undefined only when using it in a component. It is defined in the slot. Is it because in that case the translation happens prior to the content being passed to `initVue`? To clarify, here's the difference.

```
<!-- This works fine -->
<FullCalendar :options="options">
  <template #eventContent="{ event }">
    {{ $t('test') }}
  </template>
</FullCalendar>
```

```
<!-- This throws "i18n undefined" error -->

<FullCalendar :options="options">
  <template #eventContent="{ event }">
    <Event :value="event" />
  </template>
</FullCalendar>

<!-- ... -->

<Event>
  {{ $t('test') }}
</Event>
```
> I'm wondering, as a workaround, could you somehow initialize the i18n plugin globally, using Vue.use(VueI8n) (like here)?

This doesn't seem to make a difference. I updated the [repl](https://repl.it/@kgilden/fullcalendar-vue-105#components.js) to demonstrate it.

I'm not sure either how this can be solved though :shrug: . I'll think about this more.

----

PS, I don't think this is just about VueI18n. As @AbrahamBrookes mentions that this is also happening for Vuex.

**Comment by kgilden** [comment]
Now that I think about it, maybe passing the necessary props could be a fix?

```
<FullCalendar :options="options" :attributes="{ $i18n, $store }">
  <template #eventContent="{ event }">
    <Event :value="event" />
  </template>
</FullCalendar>
```

Then, initializing the Vue instance would look something like this.

```
// So the `:attributes` object would be passed in as `options' variable in here.
function initVue(initialContent: VNode[], options: ???) {
  return new Vue({
    ...options,
    props: {
      content: Array as PropType<VNode[]>
    },
    // ...
  });
}
```

I definitely feel that the prop should be separate from the `options` prop to make it clearer that these are actually not part of FullCalendar. Feel free to suggest a better name than `attributes` though.

**Comment by rangermeier** [comment]
> I'm wondering, as a workaround, could you somehow initialize the i18n plugin globally, using Vue.use(VueI8n) (like here)?

@arshaw I don't think this would suffice. You still have to pass in the same i18n instance as option when initializing the Vue instance for the event content, as shown in https://kazupon.github.io/vue-i18n/started.html#javascript

Building on the `attributes` proposal from @kgilden: Maybe it's possible to extract all non-default properties from `this.$root.$options` which would boil down to `i18n`, `store`, `router`, or whatever option is injected to the root Vue instance. That way the user wouldn't have to replicate her intial Vue setup and still could make use of her Vuex store, i18n, or router state inside the scoped slot.

**Comment by arshaw** [comment]
I've merged the PR and released everything in v5.4.0!
